### PR TITLE
FREEP-687: attachment for collections

### PR DESCRIPTION
**Pull request by neanvo**
[FREEP-687 NFT content is not showing after buying/transferring](https://cerenetwork.atlassian.net/browse/FREEP-687?atlOrigin=eyJpIjoiODlkM2U0NzE2NGQ2NDNjNDg1NjNmNjA1MjM2NTQyYzEiLCJwIjoiamlyYS1zbGFjay1pbnQifQ)

### CODE REVIEW

File: contracts/NFTAttachment.sol [hunk 1]
```diff
@@ -77,6 +84,21 @@ contract NFTAttachment is /* AccessControl, */ MetaTxContext {
...
+     *
+     * This only works with NFTs on Collections.
+     */
+    function collectionManagerAttachToNFT(uint256 nftId, bytes calldata attachment)
+    public {
+        (address collection, uint32 innerId, uint64 zeroSupply) = _parseNftId(nftId);
+        require(zeroSupply == 0, "NFT format is not compatible with Collections");
+
+        address manager = _msgSender();
+        require(Collection(collection).hasRole(COLLECTION_MANAGER_ROLE, manager), "sender is not collection manager");
```
Comment (vadim-hleif) [inline contracts/NFTAttachment.sol:97]: why do we have it? 
Is security workaround? 

Didn't get how should it help, there is still an issue - anyone with this role can change content of others NFTs, isn't it?

**Review by vadim-hleif — CHANGES_REQUESTED** [review]

File: test/collection_test.js [hunk 2]
```diff
@@ -103,6 +108,17 @@ contract("Collection", accounts => {
         assert.equal(makeOfferEv.nftId.toString(), nftId.toString());
         assert.equal(makeOfferEv.price.toNumber(), price);
 
+        await expectRevert(
```
[the remark this answers is not in the document]
Comment (neanvo) [reply test/collection_test.js:111]: I agree, tests must upgraded later, but for now its enough to have only an optimistic case (just to deploy it faster)

File: contracts/NFTAttachment.sol (hunk 1 above)
Comment (neanvo) [reply contracts/NFTAttachment.sol:97]: It's not workaround, this is implementation that looks solid.
Collection manager in the owner of the collection who can operate on it, so he is responsible for asset attachments

Comment (vadim-hleif) [reply contracts/NFTAttachment.sol:97]: > require(Collection(collection).hasRole(COLLECTION_MANAGER_ROLE, manager), "sender is not collection manager")

does this line check that specific wallet has role on specific collection? can't check hasRole method behaviour via side

**Review by vadim-hleif — APPROVED** [review]

File: contracts/Collection.sol [hunk 3]
```diff
@@ -13,14 +12,13 @@ import "./Auction.sol";
 /** This this contract describes the collection of NFTs associated with a particular user.
  *
  */
-contract Collection is OpenSeaCollection, CollectionRoyalties, CollectionIssuance, CollectionNFTAttachment, BaseTransferOperator, BaseNFT {
```
Comment (naure) [inline contracts/Collection.sol:16]: This cannot just be removed, that will not be upgradeable.

File: contracts/CollectionFactory.sol [hunk 4]
```diff
@@ -31,8 +29,6 @@ contract CollectionFactory is MetaTxContext, HasGlobalNftId {
 
     // The address of Freeport contract.
     Freeport public freeport;
-    // The address of NFTAttachment contract.
-    NFTAttachment public nftAttachment;
```
Comment (naure) [inline contracts/CollectionFactory.sol:35]: This cannot be removed either.

File: contracts/NFTAttachment.sol (hunk 1 above)
Comment (naure) [reply contracts/NFTAttachment.sol:97]: Yes.

Comment (naure) [reply contracts/NFTAttachment.sol:97]: The function is in the dependency `AccessControl`

**Pull request merged**